picard/browser/browser.py

```diff
@@ -38,18 +39,52 @@
 from picard.util import mbid_validate
 
 
-def response(code):
+def response(code, origin=None):
     if code == 200:
         resp = '200 OK'
     elif code == 400:
         resp = '400 Bad Request'
     else:
         resp = '500 Internal Server Error'
+    if origin:
+        allowOrigin = 'Access-Control-Allow-Origin: {}\r\n'.format(origin)
+    else:
+        allowOrigin = ''
     return bytearray(
         'HTTP/1.1 {}\r\n'
         'Cache-Control: max-age=0\r\n'
+        '{}'
         '\r\n'
-        'Nothing to see here.\r\n'.format(resp), 'ascii')
+        'Nothing to see here.\r\n'.format(resp, allowOrigin), 'ascii')
+
+
+def _read_origin(conn):
+    while True:
+        line = conn.readLine().data()
+        if not line or line == b'\r\n':
+            break
+        if line.startswith(b'Origin:'):
+            header, origin = line.decode().split()
+            log.debug('Browser sent origin %s', origin)
+            if _is_valid_origin(origin):
+                return origin
+            else:
+                return None
+    return None
+
+
+RE_VALID_ORIGINS = re.compile(r'(.*\.)?musicbrainz.org')
+
+
+def _is_valid_origin(origin):
+    url = urlparse(origin)
```

Comment:
`urlparse()` may raise `ValueError` in some cases:

```>>> url = urlparse('http://a.b.c[/path')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/urllib/parse.py", line 368, in urlparse
    splitresult = urlsplit(url, scheme, allow_fragments)
  File "/usr/lib/python3.6/urllib/parse.py", line 436, in urlsplit
    raise ValueError("Invalid IPv6 URL")
ValueError: Invalid IPv6 URL
```

To be safe, we should prolly catch it and return `False`.